Enable setting AWS region for terraform provisioning

rija opened the pull request:
* Update aws-instance.tf to stop hard-coding AMI id and implement AMI lookup as the AMI Id is specific to a region.
* Update terraform.tf to add aws_region as an input variable
* Update ``tf_init.sh`` to prompt operator for an AWS region and save the value in ``terraform.tfvars``
* Update the CI/CD documentation
* Update EC2 policy to enable choosing of AWS region amongst a canned list 

>Note 1: the naming convention for EIP has changed so to not hardcode region: ``eip-<application>-<environment>-<IAM Role Username>``, e.g: ``eip-gigadb-staging-Rija``

>Note 2: The EC2 policy allows use of following AWS regions:
> * ap-east-1 (Hong Kong)
> * ap-northeast-1 (Tokyo)
> * ap-southeast-1 (Singapore)
> * eu-west-3 (Paris)

rija committed:
Enable setting AWS region for terraform provisioning

* Update aws-instance.tf to stop hard-coding AMI id and implement AMI lookup as the AMI Id is specific to a region.
* Update terraform.tf to add aws_region as an input variable
* Update ``tf_init.sh`` to prompt operator for an AWS region and save the value in ``terraform.tfvars``
* Update the CI/CD documentation

rija committed:
Update EC2 policy to enable choosing of AWS region amongst a canned list

pli888 reviewed (approved):
These changes look good to me.

rija committed:
Change AMI name to be that of Centos 8

because we need to be on Centos 8 to get the recent version of openssl
that doesn't break Ansible due to the Let's Encrypt root/intermediate certificate expiration issue.

rija commented:
@rija (note to self) once rebased on ``fuw-cicd`` HEAD after #194 has been merged, apply this patch:

```
diff --git a/ops/infrastructure/terraform.tf b/ops/infrastructure/terraform.tf
index ea00f7612..aae004d7d 100644
--- a/ops/infrastructure/terraform.tf
+++ b/ops/infrastructure/terraform.tf
@@ -50,6 +50,10 @@ data "external" "callerUserName" {
   program = ["${path.module}/getIAMUserNameToJSON.sh"]
 }
 
+data "aws_availability_zones" "available" {
+  state = "available"
+}
+
 # A custom virtual private cloud network for RDS and EC2 instances
 module "vpc" {
   source  = "terraform-aws-modules/vpc/aws"
@@ -62,7 +66,7 @@ module "vpc" {
   cidr = "10.99.0.0/18"
   
   # VPC spans all the availability zones in region
-  azs = ["ap-east-1a", "ap-east-1b", "ap-east-1c"]
+  azs = data.aws_availability_zones.available.names
 
   # We can add one or more subnets into each AZ. A subnet is required to launch
   # AWS resources into a VPC and is a range of IP addresses. Each subnet has a 
```

rija committed:
Use Terraform aws_availability_zones data source to get list of AZs

We need to replace the hand listed AZs from ap-east-1 by a list automatically generated
based on the AWS region passed to the terraform script as variable.

rija committed:
Update bastion terraform module to query for AMI and update owners

the hardcoded AMI id was specific to ap-east-1, so we to work with the selected
AWS region passed as a variable input to the terraform script, we need
 to query for an AMI using aws_ami data source like we do for the dockerhost instance.

 Additionally for both instance we need to use the correct AMI owners account as indicated on
 Centos wiki: https://wiki.centos.org/Cloud/AWS

pli888 commented on ops/infrastructure/modules/aws-instance/aws-instance.tf:
```diff
@@ -56,17 +56,32 @@ resource "aws_security_group" "docker_host_sg" {
   tags = {
     Name = "gigadb_server_${var.deployment_target}_${var.owner}",
-    System = "t3_micro-centos8",
+    System = "t3_micro-centos7",
```
Should be: `System = "t3_micro-centos8",`

rija replied:
Thanks @pli888. I've fixed that.

pli888 commented on docs/SETUP_CI_CD_PIPELINE.md:
````diff
@@ -535,6 +518,18 @@ $ cd ops/infrastructure/envs/staging
+```
+  tags = {
+    Name = "gigadb_server_${var.deployment_target}_${var.owner}",
+    System = "t3_micro-centos7",
````
Should be: `System = "t3_micro-centos8",`

rija replied:
Fixed now

pli888 commented:
I've checked that this PR can be used to deploy a GigaDB website into the Tokyo Asia pacific north east region so approve this PR.

kencho51 commented:
Hi @rija I could deploy GigaDB website into `Paris` region from this PR, so I think it is good to be merged.
Just a couple of things I would like to mention:
1. Here are the steps how I provisioned the EC 2 servers:
```
% ../../../scripts/tf_init.sh --project gigascience/forks/kencho-gigadb-website --env staging
You need to specify the path to the ssh private key to use to connect to the EC2 instance: /.ssh/aws-gigadb-eu-west-3-ken.pem
You need to specify your GitLab username: kencho
You need to specify a backup file created by the files-url-updater tool: gigadbv3_20210920_v9.3.25.backup
You need to specify an AWS region: eu-west-3 
% ../../../scripts/ansible_init.sh --env staging
% terraform plan
% terramform apply
% terraform refresh
% ansible-playbook -i ../../inventories bastion_playbook.yml

TASK [Test pg_isready can connect to RDS instance] *****************************************************************************************************
fatal: [35.181.4.77]: FAILED! => {"changed": true, "cmd": ["/usr/pgsql-11/bin/pg_isready", "-h"], "delta": "0:00:00.005506", "end": "2021-10-28 07:46:43.302127", "msg": "non-zero return code", "rc": 3, "start": "2021-10-28 07:46:43.296621", "stderr": "/usr/pgsql-11/bin/pg_isready: option requires an argument -- 'h'\nTry \"pg_isready --help\" for more information.", "stderr_lines": ["/usr/pgsql-11/bin/pg_isready: option requires an argument -- 'h'", "Try \"pg_isready --help\" for more information."], "stdout": "", "stdout_lines": []}

PLAY RECAP *********************************************************************************************************************************************
35.181.4.77                : ok=5    changed=4    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0  
```
The error is because the `gigadb_db_host` in ansible.properties has not been updated, to fix this, I needed  to re-run `% ../../../scripts/ansible_init.sh --env staging` again to retrieve the latest value. The I could move on the next ansible step `% ansible-playbook -i ../../inventories dockerhost_playbook.yml`.

2. The DNS record for `staging` and `portainer.staging` needs to be updated as well, which is not mentioned in this PR.

rija commented:
Thanks @kencho51, I'll proceed with merging this branch. 

Regarding your first point, the instructions in ``docs/SETUP_CI_CD_PIPELINE.md`` was already setup to prevent you to run into that problem (see section ``4. Provision with  Terraform and perform Ansible playbook``).
Feel free to submit a patch to improve that section if it isn't clear enough. But most probably, our documentation structure may need rethinking to ensure "getting started" sections are not drown in "reference" documentation.

Regarding your second point: that is not specific to this PR, or to this task. Whenever the Elastic IP change, DNS records that were  pointing to the Elastic IP need to be updated. The documentation about it maybe absent or too basic.
I've updated the specs for #818, to add a ``docs/DNS_TLS.md`` doc to document the DNS and TLS usage and setup in our infrastrucure. 

Thanks for the suggestions.

rija committed:
Correct system tag for ec2 instance

since we are using Centos 8.*, the tags should end in centos8, not centos7

rija committed:
Correct system tag for ec2 instance in the docs

since we are using Centos 8.*, the tags should end in centos8, not centos7
